elastic/kibana
# [Synthetics] Danger highlight locations in add/edit that are not in service manifest

## Summary

Resolves #140042.

This patch fixes an issue that should not typically be a problem, but is technically possible in future versions of Kibana post-synthetics GA. If the user creates a monitor referencing a public location, and said location is removed from the manifest by us, the now-orphaned monitor will reference something that is not there.

This change will not completely resolve this problem, but it fixes the UI so it will not throw an unhandled exception and render an error boundary for the add/edit view in Synthetics UI. This is achieved by finding any location that is missing from the manifest and supplying alternative defaults. It also specifies the `danger` color to highlight that the location is misconfigured. The dead location option is also no longer available in the combo box's select. It's possible for the user to then delete the dead location from their monitor, and update the saved object via the form as usual.

## Testing instructions

0. [Configure a local synthetics service](https://docs.synthetics.elastic.dev/#/running-locally?id=setting-up-the-service-for-the-first-time). This is technically optional, but it is good to do actual e2e testing for this type of PR when reasonably practicable.

1. Create a manifest file, example below. You can easily host this by using the `http-server` package like `npx http-server --port {PORT_NUM}`
```json
{
  "throttling": {
    "download": 20,
    "upload": 10
  },
  "locations": {
    "local": {
      "url": "https://localhost:10001",
      "geo": {
        "name": "Local Synthetics Service",
        "location": {"lat": 41.25, "lon": -95.86}
      },
      "status": "ga"
    },
      "local2": {
      "url": "https://localhost:10001",
      "geo": {
        "name": "Another Local Synthetics Service",
        "location": {"lat": 41.25, "lon": -95.86}
      },
      "status": "ga"
    }
  }
}
```

2. Build this PR branch. Ensure your config is set up to point to your service, and you've set the manifest URL to the correct location as well. Some example config is included below, they should correspond to the fields specified in the service docs.

```yaml
# service keys
xpack.uptime.service.hosts: ["https://host.docker.internal:9200"]
xpack.uptime.service.devUrl: https://localhost:10001
xpack.uptime.service.username: synthetics
xpack.uptime.service.password: TmNwMBxve5qwGP5be4AxB4cffG

# manifest URL
xpack.uptime.service.manifestUrl: http://localhost:8081/manifest.json
xpack.uptime.service.tls.certificate: config/123-client.pem
xpack.uptime.service.tls.key: config/123-client.key
```

3. Start up Kibana and go to the Synthetics view (you will likely need to enable it first in Advanced Settings). Navigate to the management view and add a monitor, including at least two service locations from your self-hosted manifest.

4. Confirm you're seeing both locations represented in the overview page.

5. Kill your http server from step 1. Edit the manifest file and delete one of the locations you previously configured your monitor to reference. Restart the http server.

6. Go to the management page and attempt to edit your monitor. You should see the location highlighted in red. You could confirm that this fixes the bug in the linked issue at this point by checking out `main` and verifying that an exception is thrown.

7. At this point you've verified the issue is fixed. You could delete the old monitor from your config, save the monitor and ensure that if you come back in to edit again it is no longer showing up.

### Checklist

Delete any items that are not applicable to this PR.



### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx) |


### For maintainers
